src/DebugEngineHost/RegistryMonitor.cs

```diff
@@ -0,0 +1,148 @@
+            Security = 8,
+        }
+
+        [DllImport("advapi32.dll", SetLastError = true)]
+        private static extern int RegOpenKeyEx(IntPtr hKey, string subKey, uint options, int samDesired,
+                                               out IntPtr phkResult);
+
+        [DllImport("advapi32.dll", SetLastError = true)]
+        private static extern int RegNotifyChangeKeyValue(IntPtr hKey, bool bWatchSubtree,
+                                                          RegChangeNotifyFilter dwNotifyFilter, IntPtr hEvent,
+                                                          bool fAsynchronous);
+
+        [DllImport("advapi32.dll", SetLastError = true)]
+        private static extern int RegCloseKey(IntPtr hKey);
+
+        private const int KEY_QUERY_VALUE = 0x0001;
+        private const int KEY_NOTIFY = 0x0010;
+        private const int STANDARD_RIGHTS_READ = 0x00020000;
+
+        private static readonly IntPtr HKEY_LOCAL_MACHINE = new IntPtr(unchecked((int)0x80000002));
+
+        #endregion
+
+        private HostConfigurationSection _section;
+        private readonly bool _watchSubtree;
+
+        // Set when registry value is changed
+        private AutoResetEvent m_changeEvent;
+
+        // Set when monitoring is stopped
+        private AutoResetEvent m_stoppedEvent;
+
+        /// <summary>
+        /// Occurs when the specified registry key has changed.
+        /// </summary>
+        public event EventHandler RegChanged;
+
+        public RegistryMonitor(HostConfigurationSection section, bool watchSubtree)
+        {
+            _section = section;
+            _watchSubtree = watchSubtree;
+        }
+
+        public void Start()
+        {
+            Thread registryMonitor = new Thread(Monitor);
+            registryMonitor.IsBackground = true;
+            registryMonitor.Name = "RegistryMonitor";
+            registryMonitor.Start();
+        }
+
+        public void Stop()
+        {
+            if (m_stoppedEvent != null)
+            {
+                m_stoppedEvent.Set();
+            }
+        }
+
+        // The handle is owned by change event instance which lives while we use the handle.
+        [SuppressMessage("Microsoft.Reliability", "CA2001:AvoidCallingProblematicMethods", MessageId = "System.Runtime.InteropServices.SafeHandle.DangerousGetHandle")]
+        private void Monitor()
+        {
+            bool stopped = false;
+
+            m_stoppedEvent = new AutoResetEvent(false);
+            m_changeEvent = new AutoResetEvent(false);
+
+            IntPtr handle = m_changeEvent.SafeWaitHandle.DangerousGetHandle();
+
+            int errorCode = RegNotifyChangeKeyValue(_section.YOLOHandle, _watchSubtree, RegChangeNotifyFilter.Key | RegChangeNotifyFilter.Attribute | RegChangeNotifyFilter.Value | RegChangeNotifyFilter.Security, handle, true);
+            if (errorCode != 0) // 0 is ERROR_SUCCESS
+            {
+                throw new Win32Exception(errorCode);
```

Comment:
>throw new Win32Exception(errorCode); [](http://example.com/codeflow?start=16&length=36)

This will crash VS. I think you want to log this and go away.